dali/imgcodec/image_decoder.cc

```diff
@@ -0,0 +1,772 @@
+          if (!work.gpu_outputs.empty())
+            work.gpu_outputs[i - moved] = std::move(work.gpu_outputs[i]);
+          if (!work.temp_buffers.empty())
+            work.temp_buffers[i - moved] = std::move(work.temp_buffers[i]);
+          if (!work.rois[i])
+            work.rois[i - moved] = std::move(work.rois[i]);
+          work.sources[i - moved] = work.sources[i];
+          work.indices[i - moved] = work.indices[i];
+        }
+      } else {
+        if (fb)
+          fb->move_entry(work, i);
+        moved++;
+      }
+    }
+    if (fb)
+      fb->resize(fb->num_samples() - moved);
+}
+
+
+void ImageDecoder::copy(SampleView<GPUBackend> &out,
+                        const ConstSampleView<CPUBackend> &in,
+                        cudaStream_t stream) {
+  assert(out.shape() == in.shape());
+  assert(out.type() == in.type());
+
+  size_t nbytes = TypeTable::GetTypeInfo(out.type()).size() * volume(out.shape());
+
+  CUDA_CALL(cudaMemcpyAsync(out.raw_mutable_data(), in.raw_data(), nbytes,
+                            cudaMemcpyHostToDevice, stream));
+}
+
+std::vector<DecodeResult> ImageDecoder::Decode(DecodeContext ctx,
+                                               TensorList<CPUBackend> &out,
+                                               cspan<ImageSource *> in,
+                                               DecodeParams opts,
+                                               cspan<ROI> rois) {
+  BatchVector<SampleView<CPUBackend>> samples;
+  int N = out.num_samples();
+  samples.resize(N);
+  for (int i = 0; i < N; i++)
+    samples[i] = out[i];
+  return Decode(ctx, make_span(samples), in, opts, rois);
+}
+
+std::vector<DecodeResult> ImageDecoder::Decode(DecodeContext ctx,
+                                               TensorList<GPUBackend> &out,
+                                               cspan<ImageSource *> in,
+                                               DecodeParams opts,
+                                               cspan<ROI> rois) {
+  BatchVector<SampleView<GPUBackend>> samples;
+  int N = out.num_samples();
+  samples.resize(N);
+  for (int i = 0; i < N; i++)
+    samples[i] = out[i];
+  return Decode(ctx, make_span(samples), in, opts, rois);
+}
+
+////////////////////////////////////////////////////////////////////////////
+// Work item management
+////////////////////////////////////////////////////////////////////////////
+
+std::unique_ptr<ImageDecoder::ScheduledWork> ImageDecoder::new_work(
+    DecodeContext ctx, DecodeResultsPromise results, DecodeParams params) {
+  if (free_work_items_) {
+    std::lock_guard<std::mutex> g(work_mutex_);
+    if (free_work_items_) {
+      auto ptr = std::move(free_work_items_);
+      free_work_items_ = std::move(ptr->next);
+      ptr->results = std::move(results);
+      ptr->ctx = std::move(ctx);
+      ptr->params = std::move(params);
+      return ptr;
+    }
+  }
+
+  return std::make_unique<ScheduledWork>(ctx, results);
```

Comment:
As far as I understand we should set `.params` also for fresh work items

Reply:
Fixed.